File: Python3/757. Set Intersection Size At Least Two.py
class Solution:
    def intersectionSizeTwo(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x: (x[1], -x[0]))
        ans, p1, p2 = 0, -1, -1

        for s, e in intervals:
            # Case 1: No overlap. Current set has 0 intersection.
            # We need to add two numbers. Best greedy choice: end-1, end.
            if s > p2:
                ans += 2
                p1, p2 = e - 1, e

            # Case 2: Partial overlap. Current set has 1 intersection (p2).
            # We need to add one number. Best greedy choice: end.
            # Note: start > p1 implies start is between p1 and p2 (since start <= p2 failed above)
            elif s > p1:
                ans += 1
                p1, p2 = p2, e

            # Case 3: Full overlap (start <= p1).
            # Both p1 and p2 are in [start, end]. Do nothing.

        return ans

File: Java/757. Set Intersection Size At Least Two.java
class Solution {
    public int intersectionSizeTwo(int[][] intervals) {
        // Sort logic:
        // 1. Compare end times (a[1], b[1]). If different, sort ascending (a[1] - b[1]).
        // 2. If end times are same, compare start times. Sort descending (b[0] - a[0]).
        Arrays.sort(intervals, (a, b) -> {
            if (a[1] != b[1]) {
                return Integer.compare(a[1], b[1]);
            } else {
                return Integer.compare(b[0], a[0]);
            }
        });

        int ans = 0, p1 = -1, p2 = -1;
        for (int[] interval: intervals) {
            int s = interval[0], e = interval[1];
            if (s > p2) {
                ans += 2;
                p1 = e - 1;
                p2 = e;
            } else if (s > p1) {
                ans++;
                p1 = p2;
                p2 = e;
            }
        }

        return ans;
    }
}